mantis-publish/mantis-publish-core/src/test/java/io/mantisrx/publish/AbstractSubscriptionTrackerTest.java

```diff
@@ -0,0 +1,345 @@
+        subscriptionTracker = new TestSubscriptionTracker(archaiusConfiguration, registry, mockJobDiscovery, streamManager);
+    }
+
+    @Test
+    public void testDiscardSubscriptionsBeyondMax() {
+        config.setProperty(String.format(MAX_SUBS_PER_STREAM_FORMAT, StreamType.DEFAULT_EVENT_STREAM), 2);
+
+        streamManager.registerStream(StreamType.DEFAULT_EVENT_STREAM);
+        List<MantisServerSubscription> nextSubs = ImmutableList.of(
+                new MantisServerSubscription("id1", "select * from defaultStream where id = 1", null),
+                new MantisServerSubscription("id2", "select * from defaultStream where id = 2", null),
+                new MantisServerSubscription("id3", "select * from defaultStream where id = 3", null)
+        );
+        subscriptionTracker.setNextSubscriptions(ImmutableMap.of(StreamType.DEFAULT_EVENT_STREAM, nextSubs));
+        subscriptionTracker.refreshSubscriptions();
+
+        Set<Subscription> subs = streamManager.getStreamSubscriptions(StreamType.DEFAULT_EVENT_STREAM);
+        Set<String> subIds = subs.stream().map(Subscription::getSubscriptionId).collect(Collectors.toSet());
+        Set<String> expected = ImmutableSet.of("id1", "id2");
+        assertEquals(expected, subIds);
+    }
+
+
+    @Test
+    public void testMaxSubscriptionCountChange() {
+        config.setProperty(String.format(MAX_SUBS_PER_STREAM_FORMAT, StreamType.DEFAULT_EVENT_STREAM), 2);
+
+        streamManager.registerStream(StreamType.DEFAULT_EVENT_STREAM);
+        List<MantisServerSubscription> nextSubs = ImmutableList.of(
+                new MantisServerSubscription("id1", "select * from defaultStream where id = 1", null),
+                new MantisServerSubscription("id2", "select * from defaultStream where id = 2", null),
+                new MantisServerSubscription("id3", "select * from defaultStream where id = 3", null)
+        );
+        subscriptionTracker.setNextSubscriptions(ImmutableMap.of(StreamType.DEFAULT_EVENT_STREAM, nextSubs));
+        subscriptionTracker.refreshSubscriptions();
+
+        Set<Subscription> subs = streamManager.getStreamSubscriptions(StreamType.DEFAULT_EVENT_STREAM);
+        Set<String> subIds = subs.stream().map(Subscription::getSubscriptionId).collect(Collectors.toSet());
+        Set<String> expected = ImmutableSet.of("id1", "id2");
+        assertEquals(expected, subIds);
+
+        config.setProperty(String.format(MAX_SUBS_PER_STREAM_FORMAT, StreamType.DEFAULT_EVENT_STREAM), 4);
+
+        nextSubs = ImmutableList.of(
+                new MantisServerSubscription("id1", "select * from defaultStream where id = 1", null),
+                new MantisServerSubscription("id2", "select * from defaultStream where id = 2", null),
+                new MantisServerSubscription("id3", "select * from defaultStream where id = 3", null),
+                new MantisServerSubscription("id4", "select * from defaultStream where id = 4", null)
```

Comment:
nit: add "id5" sub to validate max cap is still valid.

Reply:
Good call.